src/Parallel/GlobalCache.hpp

```diff
@@ -620,12 +402,36 @@ template <typename Metavariables>
 template <typename GlobalCacheTag, typename Function>
 bool GlobalCache<Metavariables>::mutable_cache_item_is_ready(
     const Function& function) {
-  if (mutable_global_cache_proxy_is_set()) {
-    return Parallel::local_branch(mutable_global_cache_proxy_)
-        ->template mutable_cache_item_is_ready<GlobalCacheTag>(function);
+  using tag = MutableCacheTag<GlobalCache_detail::get_matching_mutable_tag<
+      GlobalCacheTag, Metavariables>>;
+  std::unique_ptr<Callback> optional_callback{};
+  // Reads don't need a lock
+  if constexpr (tt::is_a_v<std::unique_ptr, typename tag::tag::type>) {
+    optional_callback =
+        function(*(std::get<0>(tuples::get<tag>(mutable_global_cache_))));
   } else {
-    return mutable_global_cache_
-        ->template mutable_cache_item_is_ready<GlobalCacheTag>(function);
+    optional_callback =
+        function(std::get<0>(tuples::get<tag>(mutable_global_cache_)));
+  }
+  if (optional_callback != nullptr) {
+    // Second mutex is for vector of callbacks
+    std::mutex& mutex = tuples::get<MutexTag<tag>>(mutexes_).second;
+    {
+      // Scoped for lock guard
```

Comment:
Since this is all inside an `if` why not do:
```cpp
const std::lock_guard<std::mutex> lock(tuples::get<MutexTag<tag>>(mutexes_).second);
...
```
and avoid the extra indentation?

Reply:
Does the `return` affect this? Meaning that other destructors have to be called in addition to the lock guard destructor? Which order are they called and could that affect performance?

Reply:
destructors are called in the opposite order of constructors, and are called before returning from the function. In terms of performance, there won't be a difference between:
```cpp
}
return blah;
```
and
```cpp
return blah;
```
The machine code should be identical in those two cases.

Does that answer your questions?

Reply:
Yup! Thank you!